### DISCUSSION TITLE
Move a selection of nodes

### DISCUSSION THREAD

**Discussion by roomrys** [talmolab/sleap#1493] [Ideas]
# Problem background
- SLEAP only allows either moving one node at a time or all the nodes in the skeleton - one or all.
- Often times, groups of "nodes" := "items being tracked" will move together in the video.
- When labeling, it is cumbersome to move these nodes one by one (when it would be much easier to move the nodes as a group)

# Feature proposal
- Add an intuitive Shift + Drag in conjunction with Shift + Click to select nodes
- Store these nodes in a group/class which intercepts event handler for normal `QtNode`/`Qtinstance` `mouseMove` and `mousePress` events.
- Allow nodes from _multiple_ instances to be selected/grouped

# Implementation details
## PR 1: Add attributes/methods to store/add selected nodes.
In this PR we just want lay the groundwork by adding an attribute to `QtInstance` to store the selected nodes. Later on we will add functionality to move the selected nodes - for now, lets focus on adding/removing selected nodes.

### 1. Create a new attribute `selected_nodes` to the `QtInstance` class to store selected nodes.
We want to be able to select nodes regardless of which instance the nodes belong to, but each `QtInstance` should only store nodes that belong that particular `QtInstance`. The new attribute `selected_nodes` should:
- [ ] Be created using a [@property](https://www.programiz.com/python-programming/property) decorator.
- [ ] Be initialized to an empty list.
- [ ] Have a setter method which filters the `selected_nodes` to only those nodes within that `QtInstance` (references to the `QNode`s are stored in `QtInstance.nodes.items()`).

### 2. Modify `QNode.mousePress` method to add nodes to `QtInstance.selected_nodes` when clicked
SLEAP allows users to select entire `QtInstance`s to move at a time. This is done by setting a variable `QNode.dragParent = True` in `QNode.mousePress()` when the `Alt` key is pressed while clicking and dragging
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L1517-L1532
We want to get rid of the `QNode.dragParent` and instead populate a list of `QtInstance.selected_nodes`. The modified `QtNode.mousePress` method should:
- [ ] Add all nodes in the `QtInstance` to `QtInstance.selected_nodes` if the `Alt` key is pressed
- [ ] Append nodes that were clicked to `QtInstance.selected_nodes` if the `Shift` key is pressed

### 3. Modify `GraphicsView.mousePress` and `mouseMove` to add nodes to `QtInstance.selected_nodes` when `Shift` + drag
SLEAP allows users to zoom into an area of the video by holding `Alt` + dragging the mouse:
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L996-L999
Similarly, we want to be able to select node with a simple `Shift` + drag. To do this, we need to modify `GraphicsView.mousePress` to :
- [ ] `setDragMode(QGraphicsView.RubberBandDrag)` when the `Shift` key is pressed

Following the bit of  `GraphicsView.mouseRelease` which zooms into an selected area
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L1016-L1020
we want to add a conditional to find all nodes within the selection rectangle if `self.selection_mode = True`. The addition to `GraphicsView.mouseRelease` should:
- [ ] Find all selection items in the selection rectangle using [`GraphicsVIew.selectedItems()`](https://doc.qt.io/qt-6/qgraphicsscene.html#selectedItems).
- [ ] Filter the selection items to just `QNode`s
- [ ] Populate parent `QtInstance.selection_nodes` with `QNode`s in selection (should not overwrite current `selection_nodes`)

### 4. Modify `GraphicsView.mousePress` to remove `QtInstance.selected_nodes` 
- [ ] Remove all `selected_nodes` for each `QtInstance` in `GraphicsView.instances` if `Shift` is not pressed

### 5. Test your implementation
- [ ] Ensure that nodes can be added/removed as expected
- [ ] Ensure the correct nodes are added/removed

## PR 2: Modify methods to move `QtInstance.selected_nodes`
In this PR, we modify the existing mouse event methods to move the group of nodes stored in `QtInstance.selected_nodes`. Currently, when moving individual nodes, updating the node is handled by calling `QNode.updatePoint` from within the `QNode.mouseMove` and `mouseRelease` methods, respectively:
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L1547-L1571
As seen above, if `Alt` is held to drag the parent `QtInstance`, then the event is instead passed on to be handled by `QtInstance` (the `self.parentObject()`), and `QtInstance.updatePoints` is called instead or `QNode.updatePoint`. We want to pass all `QNode.mouseMoveEvent`s to be handled by `QtInstance.mouseMoveEvent` where the `QtInstance.selected_nodes` will be moved.

### 1. Modify `QtInstance.updatePoints` to update only points in `QtInstance.selected_nodes`
Currently `QtInstance.updatePoints` updates points for all nodes in the `QtInstance.nodes`.
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L1905-L1926
Instead, we want to modify `QtInstance.updatePoints` to 
- [ ] Only update the nodes stored in `QtInstance.selected_nodes`

### 2. Move/update nodes from `QtInstance` mouse events instead of from `QNode` mouse events
Basically, we want to only keep the functionality for when `dragParent = True` and modify this a little bit.
The `QNode.mouseMoveEvent` should be modified to:
- [ ] Pass the mouse event on to `QtInstance.mouseMoveEvent`

The `QNode.mouseReleaseEvent` should be modified to:
- [ ] Pass the mouse event on to `QtInstance.mouseReleaseEvent`

### Test implementation works
- [ ] Drag nodes and check that updates remain after switching (back to) frame

## PR 3: Modify both `QtInstance` and `QNode` `wheelEvent` to rotate selected nodes about which center?
Currently, we have a `QNode.wheelEvent` that rotates the `QtInstance`:
https://github.com/talmolab/sleap/blob/9ae2941649c81ca5b8e8301ccd99725732599fa5/sleap/gui/widgets/video.py#L1573-L1578
but, doing this with a group of selected nodes is a bit different. How would Qt know where to rotate about? Would it be easier to check the `QtInstance.selected_nodes` from a list to a reference of a new subclass of `QGraphicsObject` called `SelectedNodes`? Would the `SelectedNodes` class mess with the mouse events?

**Comment by gitttt-1234** [comment]
closing this discussion here, started a new issue to track this enhancement: https://github.com/talmolab/sleap/issues/2710
